torch_xla/experimental/pallas_kernels/multi_queries_paged_attention_kernel.py

```diff
@@ -0,0 +1,612 @@
+    for async_copy in self._async_copies:
+      async_copy.wait()
+    head_dim = self._vmem_buffer.shape[-1]
+    jax_array = self._vmem_buffer[...].astype(jnp.float32)
+    if self._scales_vmem_buffer is not None:
+      scales_jax_array = self._scales_vmem_buffer[...].astype(jnp.float32)
+    else:
+      scales_jax_array = None
+    jax_array = self._maybe_dequantize(jax_array, scales_jax_array)
+    return jax_array.reshape(-1, head_dim)
+
+
+def _flash_attention(
+    q_head_idx_per_kv,  # scalar, ranges from 0 to num_query_heads_per_kv_head
+    lengths_ref,  # [batch_size] jax.Array the length of each example
+    # input
+    q_ref,  # [1, num_q_heads_per_kv_head, num_queries_per_compute_block, head_dim]
+    k,  # [pages_per_compute_block*page_size,head_dim]
+    v,  # [pages_per_compute_block*page_size,head_dim]
+    # output
+    o_ref,  # same shape as q_ref: [1, num_q_heads_per_kv_head, num_queries_per_compute_block, head_dim]
+    l_ref,  # [1, num_q_heads_per_kv_head, num_queries_per_compute_block, MIN_BLOCK_SIZE]
+    m_ref,  # [1, num_q_heads_per_kv_head, num_queries_per_compute_block, MIN_BLOCK_SIZE]
+    l_scratch_ref,  # [num_q_heads_per_kv_head, num_queries_per_compute_block, MIN_BLOCK_SIZE]
+    m_scratch_ref,  # [num_q_heads_per_kv_head, num_queries_per_compute_block, MIN_BLOCK_SIZE]
+    acc_scratch_ref,  # [num_q_heads_per_kv_head, num_queries_per_compute_block, head_dim]
+    *,
+    num_kv_pages_per_compute_block: int,
+    num_queries_per_compute_block: int,
+    mask_value: float,
+    query_len: int,
+    page_size: int,
+    head_dim: int,
+):
+  b, kv_head_idx, q_blk_idx, kv_blk_idx = (
+      pl.program_id(0),
+      pl.program_id(1),
+      pl.program_id(2),
+      pl.program_id(3),
+  )
+
+  @pl.when(kv_blk_idx == 0)
+  def start_new_sequence():
+    l_scratch_ref[q_head_idx_per_kv] = jnp.zeros(
+        l_scratch_ref[q_head_idx_per_kv].shape, jnp.float32)
+    m_scratch_ref[q_head_idx_per_kv] = jnp.full(
+        m_scratch_ref[q_head_idx_per_kv].shape, -jnp.inf, jnp.float32)
+    acc_scratch_ref[q_head_idx_per_kv] = jnp.zeros(
+        acc_scratch_ref[q_head_idx_per_kv].shape, jnp.float32)
+
+  m_prev = m_scratch_ref[q_head_idx_per_kv]
+  l_prev = l_scratch_ref[q_head_idx_per_kv]
+  q = q_ref[0,
+            q_head_idx_per_kv, :, :].astype(jnp.float32)  # [block_q, head_dim]
+  assert q.shape == (num_queries_per_compute_block, head_dim)
+  kv_seq_len_per_kv_compute_blk = num_kv_pages_per_compute_block * page_size
+  assert k.shape == (kv_seq_len_per_kv_compute_blk, head_dim)
+  s = jnp.einsum(
+      'hd,td->ht', q, k,
+      preferred_element_type=jnp.float32)  # [block_q, block_k]
+  assert s.shape == (num_queries_per_compute_block,
+                     kv_seq_len_per_kv_compute_blk)
+
+  q_index = q_blk_idx * num_queries_per_compute_block
+  kv_index = kv_blk_idx * kv_seq_len_per_kv_compute_blk
+  kv_len = lengths_ref[b]
+  row_ids = (kv_len - query_len) + q_index + jax.lax.broadcasted_iota(
```

Comment:
Here, we assume the input query corresponds to the last (q_len) of the input kv. For example, if the input q_len is 8, and kv_len is 24, we assume the query corresponds to the kv at index [16. 24), and applies the causal mask accordingly.

@WoosukKwon please let us know if this assumption is valid or nor for the use cases in vLLM.

Reply:
Yes that's the desired behavior. Thanks for checking it out with me!